src/main/cljc/jarl/builtins/http.cljc

```diff
@@ -0,0 +1,199 @@
+
+(defn- nanos->millis [nanos]
+  (max 1 (math/round (/ nanos 1000000))))
+
+(defn parse-timeout [x]
+  (if (number? x)
+    x
+    (if (and (string? x) (utils/numeric? x))
+      (parse-long x)
+      (if-let [time-unit-pairs (re-seq #"([\.\d]+)(\p{L}+)" x)]
+        (nanos->millis (reduce + (map time-unit->nanos time-unit-pairs)))
+        (throw (errors/builtin-ex "http.send: invalid timeout value %s" x))))))
+
+(defn create-request [request]
+  (let [req  (partial get request)
+        has? (partial contains? request)]
+    (cond-> {:method               (keyword (str/lower-case (req "method")))
+             :url                  (verify-url (req "url"))
+             :follow-redirects     false
+             :force-json-decode    false
+             :force-yaml-decode    false
+             :insecure?            false
+             :raise-error          true
+             :conn-timeout         default-req-timeout}
+            (has? "enable_redirect")          (assoc :follow-redirects (req "enable_redirect"))
+            (has? "body")                     (assoc :body (to-json (req "body")))
+            (has? "raw_body")                 (assoc :body (req "raw_body"))
+            (has? "headers")                  (assoc :headers (req "headers"))
+            (has? "timeout")                  (assoc :conn-timeout (parse-timeout (req "timeout"))) ; millis here, not OPA nanos
+            (has? "force_json_decode")        (assoc :force-json-decode (req "force_json_decode"))
+            (has? "force_yaml_decode")        (assoc :force-yaml-decode (req "force_yaml_decode"))
+            (has? "raise_error")              (assoc :raise-error (req "raise_error"))
+            (has? "tls_insecure_skip_verify") (assoc :insecure? (req "tls_insecure_skip_verify")))))
+
+(defn content-type [response]
+  (when-let [ct (get-in response [:headers "content-type"])]
+    (let [stop (or (str/index-of ct ";") (count ct))]
+      (str/lower-case (subs ct 0 stop)))))
+
+(defn response-body-decoded [request response]
+  (let [ct (content-type response)]
+    (cond
+      (:force-json-decode request) (json/read-str (:body response))
+      (:force-yaml-decode request) (yaml/read-str (:body response))
+      (= ct "application/json")    (json/read-str (:body response))
+      (= ct "application/yaml")    (yaml/read-str (:body response))
+      (= ct "application/x-yaml")  (yaml/read-str (:body response))
+      :else nil)))
+
+(defn create-response [request response]
+  (let [status-code (or (:status response) (get response "status"))]
+    (if (zero? status-code)
+      ; client (currently) responsible for formatting error response correctly
+      (walk/stringify-keys response)
+      (let [resp {"status_code" status-code
+                  "status"      (status-code->status-text (:status response))
+                  "headers"     (:headers response)
+                  "body"        (response-body-decoded request response)
```

Comment:
Later, it'd be fun to see if we can optimize for this (across multiple built-ins and possible some statement types), and emit a lazy construct instead and only do work if it's actually used :).

Reply:
Do you mean like a lazy map, where we only compute the values when accessed? Yeah, I've seen some implementations of that.. could be fun for sure :)

Reply:
Yeah, something like that. A downside (?) would perhaps be that we'd not generate some errors that OPA would trigger, as we'd skip unnecessary operations.